common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java

```diff
@@ -124,47 +114,61 @@ private void tryCastSpell(SpellUnit a, SpellUnit b, SpellUnit c) {
             return;
         }
 
-        boolean isSpellInverted = isArcher;
-        List<SpellDirection> spell = Stream.of(a, b, c)
-                .map(x -> (x == SpellUnit.PRIMARY) != isSpellInverted ? SpellDirection.RIGHT : SpellDirection.LEFT)
-                .toList();
-        for (int i = 0; i < spellInProgress.length; ++i) {
-            if (spellInProgress[i] != spell.get(i)) {
-                sendCancelReason(Component.translatable("feature.wynntils.quickCast.incompatibleInProgress"));
-                return;
-            }
+        Spell spell = new Spell(a, b, Models.Character.getClassType() == ClassType.ARCHER);
+        if (spells.contains(spell)) {
+            return;
         }
 
+        spells.offer(spell);
         lastSelectedSlot = McUtils.inventory().selected;
-        List<SpellDirection> remainder = spell.subList(spellInProgress.length, spell.size());
-        remainder.stream().map(SpellDirection::getSendPacketRunnable).forEach(SPELL_PACKET_QUEUE::add);
     }
 
     @SubscribeEvent
     public void onTick(TickEvent e) {
-        if (!Models.WorldState.onWorld()) return;
-
-        // Clear spell after the 40 tick timeout period
-        if (spellCountdown > 0 && --spellCountdown <= 0) spellInProgress = SpellDirection.NO_SPELL;
-
-        if (SPELL_PACKET_QUEUE.isEmpty()) return;
         if (--packetCountdown > 0) return;
+        if (currentSpell.isEmpty() && !pollSpell()) return;
 
         int currSelectedSlot = McUtils.inventory().selected;
         boolean slotChanged = currSelectedSlot != lastSelectedSlot;
-
         if (slotChanged) McUtils.sendPacket(new ServerboundSetCarriedItemPacket(lastSelectedSlot));
-        SPELL_PACKET_QUEUE.poll().run();
+        if (safeCasting.get()) {
+            currentSpell.poll().getSendPacketRunnable().run();
+        } else {
+            do {
+                currentSpell.peek().getSendPacketRunnable().run();
+            } while (currentSpell.poll() == SpellDirection.LEFT && !SPELL_PACKET_QUEUE.isEmpty());
+            while (currentSpell.peek() == SpellDirection.LEFT) {
+                currentSpell.poll().getSendPacketRunnable().run();
+            }
+        }
         if (slotChanged) McUtils.sendPacket(new ServerboundSetCarriedItemPacket(currSelectedSlot));
 
         // Waiting a few ticks is useful for avoiding lag related input-overlaps
-        if (!SPELL_PACKET_QUEUE.isEmpty()) packetCountdown = 3;
+        packetCountdown = Math.max(1, rightClickCooldown.get());
+    }
+
+    private boolean pollSpell() {
+        spells.remove(spell);
+        spell = null;
```

Comment:
Overall this whole method seems really weird. You return early in multiple cases, but still do this before returning, every time..? It could also use a bit of spacing in logical places.

Reply:
`spell` is the currently active spell. `pollSpell` gets called once the spell is finished casting. than and only than can I remove the spell from the queue because I don't what It to get In the queue while its still active because when the user releases the appropriate spell button it would feel unresponsive when the spell casts again.

Reply:
The problem is that you have to explain it to me. Now I see, but anyone reading this code later won't. Neither `spell` or `pollSpell` is something I remotely understand when reading, which is a problem, they need better names. As for the method, again, it is not something that is easy to understand when looking at/reading, which should be the goal of every method.

Reply:
is it fine now?